[GLIB] Add API to control originStorageRatio and totalStorageRatio

pgorszkowski-igalia opened the pull request:
#### 34c3ebb7bacda0a74abfd5c7fe61e2cc387dba4b
<pre>
[GLIB] Add API to control originStorageRatio and totalStorageRatio
<a href="https://bugs.webkit.org/show_bug.cgi?id=254905">https://bugs.webkit.org/show_bug.cgi?id=254905</a>

Reviewed by Carlos Garcia Campos.

Add API to control originStorageRatio and totalStorageRatio.

* Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.cpp:
(webkitWebsiteDataManagerSetProperty):
(webkit_website_data_manager_class_init):
(webkitWebsiteDataManagerGetDataStore):

Canonical link: <a href="https://commits.webkit.org/265317@main">https://commits.webkit.org/265317@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/114743677191e106831fff9ff1af6f4ae30530c7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10606 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10828 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11108 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12253 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10177 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/10621 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13200 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10796 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13101 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10767 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11690 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8917 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12654 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8984 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9575 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16835 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10055 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9725 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12978 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10195 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8276 "3 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9349 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13613 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1180 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10051 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.cpp:
```diff
@@ -486,6 +492,25 @@ static void webkit_website_data_manager_class_init(WebKitWebsiteDataManagerClass
             nullptr, nullptr,
             FALSE,
             static_cast<GParamFlags>(WEBKIT_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY)));
+
```
```suggestion
```

mcatanzaro replied:
One problem here is that [WebKit's GB units are incorrect, and are actually GiB](https://bugs.webkit.org/show_bug.cgi?id=209314). This is really confusing. If the default quota is defined using WTF::GB then it's really GiB and we need to document that accordingly. :/

pgorszkowski-igalia replied:
I will create an issue about that and fix it.

pgorszkowski-igalia replied:
https://bugs.webkit.org/show_bug.cgi?id=258342

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.cpp:
```diff
@@ -486,6 +492,25 @@ static void webkit_website_data_manager_class_init(WebKitWebsiteDataManagerClass
+     * Max size of origin data storage. This includes CacheStorage and IndexedDB.
+     * Value of 0 means no change so whatever the default value is it will be used.
+     *
+     * Since: 2.40
```
```suggestion
     * Since: 2.42
```

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.cpp:
```diff
@@ -118,6 +119,8 @@ struct _WebKitWebsiteDataManagerPrivate {
     GUniquePtr<char> domCacheDirectory;
 #endif
+
+    guint64 perOriginStorageQuota { 0 };
```
```suggestion
    guint64 perOriginStorageQuota;
```

(The priv struct is already zero-initialized.)

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.cpp:
```diff
@@ -486,6 +492,25 @@ static void webkit_website_data_manager_class_init(WebKitWebsiteDataManagerClass
+    /**
+     * WebKitWebsiteDataManager:per-origin-storage-quota:
+     *
+     * Max size of origin data storage. This includes CacheStorage and IndexedDB.
```
I would not mention CacheStorage and IndexedDB here because this will become obsolete pretty quick as more data types are added to the origin storage manager:

 * I think the limit already applies to filesystem and localstorage
 * But not CacheStorage, because CacheStorage is only managed by OriginStorageManager if using UnifiedOriginStorageLevel::Standard, but that's currently only used for development. We unconditionally use UnifiedOriginStorageLevel::Basic to match what Apple does. So I think this limit won't apply to it yet?

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.cpp:
```diff
@@ -486,6 +492,25 @@ static void webkit_website_data_manager_class_init(WebKitWebsiteDataManagerClass
+        PROP_PER_ORIGIN_STORAGE_QUOTA,
+        g_param_spec_uint64("per-origin-storage-quota",
+            _("Per origin storage quota"),
+            _("The maximum size of storage per origin (CacheStorage and IndexedDB)"),
```
Ditto regarding not mentioning the particular data types managed by this property

mcatanzaro reviewed (commented):
Also it needs some basic API test (can be very simple, like just creating the WebKitWebsiteDataManager using this property and verify that it doesn't fail).

mcatanzaro reviewed (commented):
Looks good.

(Since this adds new API, a second reviewer is required to approve.)

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.cpp:
```diff
@@ -486,6 +492,24 @@ static void webkit_website_data_manager_class_init(WebKitWebsiteDataManagerClass
+        PROP_PER_ORIGIN_STORAGE_QUOTA,
+        g_param_spec_uint64("per-origin-storage-quota",
+            _("Per origin storage quota"),
+            _("The maximum size of storage per origin"),
```
We don't mark these for translations anymore, actually we don't use them anymore, just pass nullptr.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.cpp:
```diff
@@ -486,6 +492,24 @@ static void webkit_website_data_manager_class_init(WebKitWebsiteDataManagerClass
+     * WebKitWebsiteDataManager:per-origin-storage-quota:
+     *
+     * Max size of origin data storage.
+     * Value of 0 means no change so whatever the default value is it will be used.
```
We should document the units, is this size in bytes? KB? MB?

pgorszkowski-igalia replied:
It is in bytes (the same as defaultStandardReportedQuota: https://github.com/WebKit/WebKit/blob/main/Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp#L72)

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.cpp:
```diff
@@ -486,6 +492,24 @@ static void webkit_website_data_manager_class_init(WebKitWebsiteDataManagerClass
             static_cast<GParamFlags>(WEBKIT_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY)));
+
+    /**
+     * WebKitWebsiteDataManager:per-origin-storage-quota:
```
I guess this only affects local storage, right? not indexeddb or any other storage. I would clarify this in the property name or the documentation at least. Maybe max-local-storage-size-per-origin or something similar

mcatanzaro replied:
No, it affects everything using the origin storage manager, which may change in the future. But see [here](https://github.com/WebKit/WebKit/pull/12306#discussion_r1155991217).

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.cpp:
```diff
@@ -486,6 +497,38 @@ static void webkit_website_data_manager_class_init(WebKitWebsiteDataManagerClass
+    /**
+     * WebKitWebsiteDataManager:origin-storage-ratio:
+     *
+     * The ratio(0-1) which is used to calculate the origin data storage base on disk space.
```
```suggestion
     * The ratio (0-1) which is used to calculate the origin data storage base on disk space.
```

We need a little more detail in this comment. 0.1 would mean the database is allowed to consume 10% of the user's disk space, correct? So values should be very small, e.g. 0.02 would be a reasonable default.

What happens if you do not change the default value and leave it at 0? This should be documented too. I doubt it means "no limit," does it?

pgorszkowski-igalia replied:
> What happens if you do not change the default value and leave it at 0? This should be documented too. I doubt it means "no limit," does it?

The default value 0 means "no change so whatever the default value is it will be used.", I will add this information to documentation

> 0.1 would mean the database is allowed to consume 10% of the user's disk space, correct?

yes

pgorszkowski-igalia replied:
One more note: If we would set the ratio to 0.0 then quota be 0, so nothing would be cached/stored.

So I limit the range of the ratio to (0.0, 1.0>

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.cpp:
```diff
@@ -90,7 +90,9 @@ enum {
-    PROP_IS_EPHEMERAL
+    PROP_IS_EPHEMERAL,
+    PROP_PER_ORIGIN_STORAGE_QUOTA,
+    PROP_ORIGIN_STORAGE_RATIO
```
Hm, I thought they were two different ways of configuring the same setting, but your naming indicates that the per-origin storage quota applies to particular origins, while the origin storage ratio applies to the entire database. Is that correct? We should be very clear about this in the documentation.

pgorszkowski-igalia replied:
There are:
 * totalQuotaRatio
 * originStorageRatio
 * perOriginStorageQuota

originStorageRatio(not specified by default) if is specified it is used instead of perOriginStorageQuota(which is by default 10GB).

totalQuotaRatio is used for whole.

pgorszkowski-igalia replied:
And currently there is no API to control totalQuotaRatio

pgorszkowski-igalia replied:
Should we add also API for totalQuotaRatio?

mcatanzaro replied:
I suppose. But maybe we'd better talk to @szewai to make sure all of these quotas are likely to remain supported in the future.

szewai replied:
There are two types of quota now: 
1. Origin quota -- how much can an origin store; if it is exceeded, request will fail.
2. Total quota -- how much can a dataStore (all origins) store; if it is exceeded, eviction will happen.

Origin quota can be configured in two ways:
- Setting value directly -- `perOriginStorageQuota`. This is the existing way of configuring quota.
- Calculating value based on hard disk size -- `originQuotaRatio`: quota = originQuotaRatio * totalDiskSpace. If `originQuotaRatio` is set, calculation will be based on ratio, no matter perOriginStorageQuota is set or not.

Total quota can be configured in one way:
- Calculating value on hard disk size -- `totalQuotaRatio`: quota = totalQuotaRatio * totalDiskSpace. If `totalQuotaRatio` is not set, there is no eviction based on usage.

The motivation for new quota mechanism is to be compatible with other browser engines (https://developer.mozilla.org/en-US/docs/Web/API/Storage_API/Storage_quotas_and_eviction_criteria#how_much_data_can_be_stored). 

I just enabled "calculating quotas based on total disk space" on Cocoa platforms: https://bugs.webkit.org/show_bug.cgi?id=255095. I think we might deprecate `perOriginStorageQuota` if it's not in use anywhere.

mcatanzaro replied:
> I think we might deprecate perOriginStorageQuota if it's not in use anywhere.

Well we don't want to expose it as public API then. Good thing I CCed you.

pgorszkowski-igalia replied:
@mcatanzaro: does it mean that this patch should be closed without merging

pgorszkowski-igalia replied:
Or maybe just add in this PR API for totalQuotaRatio(and remove: perOriginStorageQuota and originQuotaRatio)?

mcatanzaro replied:
If totalQuotaRatio is all you need, then I would add just that one. Let's expose only what you actually need.

pgorszkowski-igalia replied:
Currently the most important from our point of view is perOriginStorageQuota

pgorszkowski-igalia replied:
@mcatanzaro : what can we do with this PR?

mcatanzaro replied:
Based on https://github.com/WebKit/WebKit/pull/12306#discussion_r1174153785 I guess we should expose `originQuotaRatio` and `totalQuotaRatio` rather than `perOriginStorageQuota`.

pgorszkowski-igalia replied:
ok, so I removed API for perOriginStorageQuota and added API for totalQuotaRatio

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.cpp:
```diff
@@ -479,6 +490,41 @@ static void webkit_website_data_manager_class_init(WebKitWebsiteDataManagerClass
+    /**
+     * WebKitWebsiteDataManager:origin-storage-ratio:
+     *
+     * The ratio which is used to calculate the origin data storage base on disk space.
```
```suggestion
     * The ratio which is used to calculate the origin data storage based on disk space.
```

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.cpp:
```diff
@@ -479,6 +490,41 @@ static void webkit_website_data_manager_class_init(WebKitWebsiteDataManagerClass
+     *
+     * The ratio which is used to calculate the origin data storage base on disk space.
+     * This value will range from 0.0 to 1.0.
+     * Value of 0.0 means no change.
```
No change relative to what? The documentation needs to be clear enough for developers to understand without needing to look at the code of OriginStorageManager, and it's a construct-only property so inherently it cannot change.

I see what you mean is "if 0, don't call `WebsiteDataStoreConfiguration::setOriginQuotaRatio`" but that's an implementation detail. What we need to document is what is the origin quota ratio going to be if left at 0.0? Perhaps that should mean "no storage allowed" as if you wanted there to be no quota at all such that an origin could fill up your hard disk, you would set it to 1.0.

We also need to document precisely how the ratio works. Does it limit the amount of data that can be stored relative to available disk space? Or does it limit what can be stored relative to the total storage ratio?

mcatanzaro reviewed (changes requested):
The documentation will need to be more clear about what exactly the properties do.

We'll probably want to think about how to name these new properties. `origin-storage-ratio` and `total-storage-ratio` is not very clear.

I wonder if we could do a better API test for this somehow to make sure it actually works, without the result depending on available disk space. That might be tricky, but if there properties were to change such that setting a 0.0 ratio were to indicate that no storage is allowed, then we could at least verify that setting the ratios to 0.0 blocks website data storage.

@carlosgcampos we'll need you to review this too.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.cpp:
```diff
@@ -479,6 +490,45 @@ static void webkit_website_data_manager_class_init(WebKitWebsiteDataManagerClass
+     * The origin data storage if it is exceeded, request will fail.
+     * This value has to be in range from 0.0 to 1.0.
+     * Value of 0.0 means that data storage is not allowed.
+     * Value out of the range 0.0 - 1.0 means nothing will change.
```
Now the documentation says the value has to be in the range 0.0 to 1.0, but in fact it can be -1.0 to 1.0. I don't think this is less confusing.

pgorszkowski-igalia replied:
I had to extend the range to give a possibility to set some default value which does not change anything :( I could not use 1.0 as default because in that case the behaviour is different in contrast to not set at all. Do you have any suggestion here?

mcatanzaro replied:
What I'm trying to suggest is that "default value which does not change anything" is meaningless to API users. Who knows what that is? (I don't. :) So forget about it. Always call setOriginQuotaRatio and setTotalQuotaRatio. The only default value should be the one that you pass to g_param_spec_double().

mcatanzaro replied:
(That said, I'd be curious what the internal WebCore defaults are for purposes of choosing what our defaults should be.)

pgorszkowski-igalia replied:
OriginQuotaRatio and TotalQuotaRatio are optional. If we always call setOriginQuotaRatio and setTotalQuotaRatio we change the default behavior of QuotaManager(NetworkStorageManager::originStorageManager in case of originQuotaRatio) which will use this value (in contrast to use default quota value (1000MB) or the one set by perOriginStorageQuota)

pgorszkowski-igalia replied:
In case of TotalQuotaRatio if we don't set it (the default behaviour) then there won't be any checking and eventual eviction done(NetworkStorageManager::spaceGrantedForOrigin)

pgorszkowski-igalia replied:
There are functions which reset the default value to make possible to fallback to old functionality:
WKWebsiteDataStoreConfigurationClearOriginQuotaRatio
WKWebsiteDataStoreConfigurationClearTotalQuotaRatio

and in current Mac implementation the default values are:
https://github.com/WebKit/WebKit/blob/main/Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm#L66
static constexpr double defaultBrowserTotalQuotaRatio = 0.8;
static constexpr double defaultBrowserOriginQuotaRatio = 0.6;

mcatanzaro replied:
I'm going to resolve this discussion thread because Carlos has asked you to use -1 for no change after all. I'll drop my complaint.

mcatanzaro commented on Tools/TestWebKitAPI/glib/WebKitGLib/WebViewTest.cpp:
```diff
@@ -273,6 +273,15 @@ void WebViewTest::assertFileIsCreated(const char *filename)
     g_assert_true(g_file_test(filename, G_FILE_TEST_EXISTS));
 }
 
+void WebViewTest::assertFileIsNotCreated(const char *filename)
```
```suggestion
void WebViewTest::assertFileIsNotCreated(const char* filename)
```

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.cpp:
```diff
@@ -479,6 +490,41 @@ static void webkit_website_data_manager_class_init(WebKitWebsiteDataManagerClass
+     *
+     * The ratio which is used to calculate the origin data storage based on disk space.
+     * The origin data storage if it is exceeded, request will fail.
+     * Value of 0.0 means that data storage is not allowed.
```
So, if I understand this correctly, this is the percentage of the disk space (actually the partition/volume) that could be used to store data for every origin, right? I find the documentation a bit confusing.

pgorszkowski-igalia replied:
Yes, in case originStorageRatio it is a percentage of disk space PER ORIGIN, in case of totalStorageRatio it is a percentage of disk space for all origins.
The difference between origin and total is also how browser reacts in case of exceeding:
* in case of origin it fails with request
* in case of total it starts eviction of storage

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.cpp:
```diff
@@ -479,6 +490,41 @@ static void webkit_website_data_manager_class_init(WebKitWebsiteDataManagerClass
+        g_param_spec_double("origin-storage-ratio",
+            nullptr, nullptr,
+            0.0, 1.0, 0.6,
+            static_cast<GParamFlags>(WEBKIT_PARAM_WRITABLE | G_PARAM_CONSTRUCT_ONLY)));
```
Since this is a construct property it means it will always be set to the default value on construction. I think we should keep current behavior when these properties are not set. Since we don't have optional, I think we can use -1 as the default behavior. So, I would use -1 as the min and default values, and 1 as the maximum.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.cpp:
```diff
@@ -506,6 +552,12 @@ WebKit::WebsiteDataStore& webkitWebsiteDataManagerGetDataStore(WebKitWebsiteData
             configuration->setCacheStorageDirectory(FileSystem::stringFromFileSystemRepresentation(priv->domCacheDirectory.get()));
 #endif
+        if (priv->originStorageRatio >= 0.0 && priv->originStorageRatio <= 1.0)
+            configuration->setOriginQuotaRatio(priv->originStorageRatio);
```
With the current values this is always true, since 0 is the min and 1 the maximum. If we change to use -1, here we would only need to check if != -1

pgorszkowski-igalia replied:
Maybe better would be check if >=0? So everything <0 will be ignored

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.cpp:
```diff
@@ -1637,3 +1689,38 @@ GList* webkit_website_data_manager_get_itp_summary_finish(WebKitWebsiteDataManag
+ *
+ * Reset existing origin storage ratio.
+ *
+ * Resetting origin storage ratio causes that quota is based on perOriginStorageQuota.
```
This is confusing, what is `perOriginStorageQuota`? I don't think we want functions to reset the value, it's a construct only property, it shouldn't change. I think resetting is probably needed for WKTR to ensure changing this in one tests don't affect others, but I don't think we need this in the public API.

pgorszkowski-igalia replied:
You are right, I will remove these functions for clearing.

carlosgcampos reviewed (changes requested):
Properties look good, but documentation need to be clearer. And I think we should ensure we don't change the current behavior is the properties are not explicitly set.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.cpp:
```diff
@@ -479,6 +490,42 @@ static void webkit_website_data_manager_class_init(WebKitWebsiteDataManagerClass
+    /**
+     * WebKitWebsiteDataManager:origin-storage-ratio:
+     *
+     * The ratio which is used to calculate the origin data storage based on disk space.
```
I still think we need to explain this better. If I've understood it correctly this value is the percentage of volume space that can be used for data storage for every domain. Right? Because the ration to calculate makes me wonder, but how is it calculated, how can I decide a value if I don't know how it's calculated?

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.cpp:
```diff
@@ -479,6 +490,42 @@ static void webkit_website_data_manager_class_init(WebKitWebsiteDataManagerClass
+     * WebKitWebsiteDataManager:origin-storage-ratio:
+     *
+     * The ratio which is used to calculate the origin data storage based on disk space.
+     * The origin data storage if it is exceeded, request will fail.
```
What request would fail? a storage request? or a load request or what? How does it fail? the load fails? or a console message is shown?

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.cpp:
```diff
@@ -479,6 +490,42 @@ static void webkit_website_data_manager_class_init(WebKitWebsiteDataManagerClass
+     *
+     * The ratio which is used to calculate the origin data storage based on disk space.
+     * The origin data storage if it is exceeded, request will fail.
+     * Values < 0.0 will be ignored.
```
-1 is < 0 and it's not ignored, means use the default WebKit value.

pgorszkowski-igalia replied:
If we not call setOriginQuotaRatio or setTotalQuotaRatio then they stay as null - this is what I mean by ignore. What is your proposal for comment about -1 value? 

The default values for ratios (origin and total) is that they are not set (Optional is null) which means:
1. in case of totalStorageRatio it means that not checking will be done at all (and any eviction in consequence)
2. in case of originStorageRatio it means that different value is used to calculate quota limitation (perOriginStorageQuota or default 1GB if perOriginStorageQuota was not set)

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.cpp:
```diff
@@ -506,6 +553,12 @@ WebKit::WebsiteDataStore& webkitWebsiteDataManagerGetDataStore(WebKitWebsiteData
         if (priv->domCacheDirectory)
             configuration->setCacheStorageDirectory(FileSystem::stringFromFileSystemRepresentation(priv->domCacheDirectory.get()));
 #endif
+        if (priv->originStorageRatio >= 0.0)
```
I'm fine doing this, but the minimum value is -1 so we could compare to -1. In any case, the documentation should say -1 and not any negative value.

carlosgcampos reviewed (commented):
The patch looks good to me, only the documentation needs some more clarification.

pgorszkowski-igalia committed:
[GLIB] Add API to control originStorageRatio and totalStorageRatio
https://bugs.webkit.org/show_bug.cgi?id=254905

Reviewed by Carlos Garcia Campos.

Add API to control originStorageRatio and totalStorageRatio.

* Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.cpp:
(webkitWebsiteDataManagerSetProperty):
(webkit_website_data_manager_class_init):
(webkitWebsiteDataManagerGetDataStore):

Canonical link: https://commits.webkit.org/265317@main

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.cpp:
```diff
@@ -479,6 +490,42 @@ static void webkit_website_data_manager_class_init(WebKitWebsiteDataManagerClass
+     * The percentage of volume space that can be used for data storage for every domain.
+     * If the maximum storage is reached the storage request will fail with a QuotaExceededError exception.
+     * A value of 0.0 means that data storage is not allowed. A value of -1.0, which is the default,
+     * means WebKit will use the default quota (1GB).
```
Where is this value in the code? It needs a comment to update the API docs here if it ever changes in the future.

Also, it should be `1 GB` with a space, rather than `1GB`.

(This really is 1 GB and not 1 GiB, correct?)
